### PR TITLE
[vwweconnect] [WIP] Initial contribution

### PR DESCRIPTION
**Pull request by jannegpriv**
[vwweconnect] [WIP] Initial contribution.

My goal is to add support for the VW We Connect portal to openHAB.

Link to community thread with URL to latest built jar-file:
https://community.openhab.org/t/new-binding-volkswagen-we-connect/92057

### CODE REVIEW

**Review by fwolter — CHANGES_REQUESTED** [review]
Can you add yourself to the CODEOWNERS file?

You need to add your binding to bundles/pom.xml and bom/openhab-addons/pom.xml.

**Comment by jannegpriv** [comment]
@Hilbrand I've changed my upstream to point to openhab-addons after the name change but I fail to merge into my 2,5,x branch:

```
$ git remote add upstream https://github.com/openhab/openhab-addons.git
jannegpriv in /Users/jannegpriv/git/openhab-master/git/openhab-addons (2.5.x) (22 entries, 6 hidden)
$ git remote -v
origin	https://github.com/jannegpriv/openhab2-addons.git (fetch)
origin	https://github.com/jannegpriv/openhab2-addons.git (push)
upstream	https://github.com/openhab/openhab-addons.git (fetch)
upstream	https://github.com/openhab/openhab-addons.git (push)
$ git fetch upstream
remote: Enumerating objects: 1539, done.
remote: Counting objects: 100% (1539/1539), done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 19016 (delta 1533), reused 1533 (delta 1533), pack-reused 17477
Receiving objects: 100% (19016/19016), 8.55 MiB | 2.77 MiB/s, done.
Resolving deltas: 100% (6773/6773), completed with 601 local objects.
From https://github.com/openhab/openhab-addons
 * [new branch]            2.5.x      -> upstream/2.5.x
 * [new branch]            dependabot/maven/bundles/org.openhab.persistence.dynamodb/com.fasterxml.jackson.core-jackson-databind-2.9.10.5 -> upstream/dependabot/maven/bundles/org.openhab.persistence.dynamodb/com.fasterxml.jackson.core-jackson-databind-2.9.10.5
 * [new branch]            dependabot/maven/bundles/org.openhab.persistence.jdbc/mysql-mysql-connector-java-8.0.16 -> upstream/dependabot/maven/bundles/org.openhab.persistence.jdbc/mysql-mysql-connector-java-8.0.16
 * [new branch]            main       -> upstream/main
 * [new tag]               2.5.0      -> 2.5.0
 * [new tag]               V0.6.0     -> V0.6.0
 * [new tag]               v0.4.0     -> v0.4.0
 * [new tag]               v0.5.0     -> v0.5.0
jannegpriv in /Users/jannegpriv/git/openhab-master/git/openhab-addons (2.5.x) (22 entries, 6 hidden)
$ git merge upstream/2.5.x
fatal: refusing to merge unrelated histories
```

Do I have to fork and clone the openhab-addons from start?

**Comment by fwolter** [comment]
It might be easier to start over. Nevertheless, you need to migrate your code manually to OH3 or use the migrated code by the openhab-bot.

**Comment by jannegpriv** [comment]
> It might be easier to start over. Nevertheless, you need to migrate your code manually to OH3 or use the migrated code by the openhab-bot.

By start-over you mean to close this PR, do a new fork/clone and adapt my binding to 3.0 and then open a new PR towards 3.0?
Where can I get more info about this openhab-bot?

**Comment by fwolter** [comment]
Correct. The procedure is described here #8512. The bot posted a link to the migrated code above: https://github.com/openhab/openhab-addons/pull/8314#issuecomment-695852654

**Comment by jannegpriv** [comment]
> Correct. The procedure is described here #8512. The bot posted a link to the migrated code above: [#8314 (comment)](https://github.com/openhab/openhab-addons/pull/8314#issuecomment-695852654)

OK, thanks for the tips! :-)

**Comment by jannegpriv** [comment]
> Correct. The procedure is described here #8512. The bot posted a link to the migrated code above: [#8314 (comment)](https://github.com/openhab/openhab-addons/pull/8314#issuecomment-695852654)

@fwolter @Hilbrand  
I managed to get an openHAB 3.0 eclipse environment working and the VWWeconnect binding builds fine.
However when I want to debug the binding it has some depenedencies listed in its pom.xml:

```
<dependency>
      <groupId>org.jsoup</groupId>
      <artifactId>jsoup</artifactId>
      <version>1.8.3</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>com.jayway.jsonpath</groupId>
      <artifactId>json-path</artifactId>
      <version>2.4.0</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.ow2.asm</groupId>
      <artifactId>asm</artifactId>
      <version>5.0.4</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>net.minidev</groupId>
      <artifactId>accessors-smart</artifactId>
      <version>1.2</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>net.minidev</groupId>
      <artifactId>json-smart</artifactId>
      <version>2.3</version>
      <scope>compile</scope>
    </dependency>
```

I've added the binding to the `Run Requirements` of `app.bndrun` and when  pressing `Resolve` button I get the following unresolved requirements:

```
Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.openhab.binding.vwweconnect)
          ⇒ [org.openhab.binding.vwweconnect version=3.0.0.202010300859]
              ⇒ osgi.wiring.package: (&(osgi.wiring.package=com.jayway.jsonpath)(version>=2.4.0)(!(version>=3.0.0)))
                  ⇒ [com.jayway.jsonpath.json-path version=2.4.0]
                      ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.minidev.json.writer)(version>=2.3.0)(!(version>=3.0.0)))
                          ⇒ [net.minidev.json-smart version=2.3.0]
                              ⇒ osgi.wiring.package: (&(osgi.wiring.package=net.minidev.asm)(version>=1.2.0)(!(version>=2.0.0)))
                                  ⇒ [net.minidev.accessors-smart version=1.2.0]
                                      ⇒ osgi.wiring.package: (&(osgi.wiring.package=org.objectweb.asm)(version>=5.0.0)(!(version>=6.0.0)))
```

I'm not sure this is a new problem or if I had it on my old build environment for 2.5.x and resolved it there in a way I cannot remember now, do you have any tips how to fix this?

**Comment by fwolter** [comment]
The groupId `org.objectweb` of `asm` doesn't seem to match your maven dependency.

**Comment by jannegpriv** [comment]
> The groupId `org.objectweb` of `asm` doesn't seem to match your maven dependency.

After doing some `mvn clean install` on both binding and openhab-distro and refreshing Eclipse the `Resolve` finally succeeded.  

When debugging and starting a scan from the new GUI I got the following WARN:
```
Representation property 'WVABGDJFGJKSHDUGHJ' of discovery result for thing 'vwweconnect:vehicle:3f65494f1c:WVABGDJFGJKSHDUGHJ' is missing in properties map. It has to be fixed by the bindings developer.
```

That is a new warning, how do I fix that?

I got the following definitions in the vehicle.xml:

```
               ...
                <properties>
			<property name="vendor">Volkswagen</property>
		</properties>

		<representation-property>vin</representation-property>

		<config-description>
			<parameter name="vin" type="text" required="true">
				<label>Vehicle Identification Number</label>
				<description>VIN of the vehicle associated with this Thing</description>
			</parameter>
		</config-description>
	</thing-type>

```

What is the `properties map` ?

File: bundles/org.openhab.binding.vwweconnect/src/main/java/org/openhab/binding/vwweconnect/internal/discovery/VWWeConnectDiscoveryService.java
Comment (fwolter) [inline bundles/org.openhab.binding.vwweconnect/src/main/java/org/openhab/binding/vwweconnect/internal/discovery/VWWeConnectDiscoveryService.java:84]: The representation property must be the name of the property, not the value.

**Review by fwolter — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.vwweconnect/src/main/java/org/openhab/binding/vwweconnect/internal/discovery/VWWeConnectDiscoveryService.java
Comment (jannegpriv) [reply bundles/org.openhab.binding.vwweconnect/src/main/java/org/openhab/binding/vwweconnect/internal/discovery/VWWeConnectDiscoveryService.java:84]: aha :-)

Thanks for the info!

I've now created a new PR #8934 for openHAB3.0, this PR can be closed.

**Pull request closed without merging**
